library/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/processor/MutationResolver.kt

```diff
@@ -11,26 +11,179 @@ import java.util.LinkedList
+
+                    if (nextNewEntry is Entry.Reference &&
+                        nextOldEntry is Entry.Reference &&
+                        nextOldEntry.id == nextNewEntry.id
+                    ) {
+                        na[i + 1] = Entry.Index(entry.index + 1)
+                        oa[entry.index + 1] = Entry.Index(i + 1)
+                    }
+                }
+            }
+        }
+
+        // 5th pass
+        // Similar to 4th pass, except it processes entries in descending order.
+        if (na.size > 1) {
+            for (i in (1 until (na.size - 1)).reversed()) {
+                val entry = na[i]
+                if (entry is Entry.Index && entry.index - 1 >= 0) {
+                    val prevNewEntry = na[i - 1]
+                    val prevOldEntry = oa[entry.index - 1]
+                    if (prevNewEntry is Entry.Reference &&
+                        prevOldEntry is Entry.Reference &&
+                        prevOldEntry.id == prevNewEntry.id
+                    ) {
+                        na[i - 1] = Entry.Index(entry.index - 1)
+                        oa[entry.index - 1] = Entry.Index(i - 1)
+                    }
+                }
+            }
+        }
+
+        // Final pass
+        // Constructing the actual diff from information stored in `oa` and `na`.
         val updates: MutableList<MobileSegment.WireframeUpdateMutation> = LinkedList()
         val adds: MutableList<MobileSegment.Add> = LinkedList()
-        var prevWireframe: MobileSegment.Wireframe? = null
-        currentSnapshot.forEach { currentWireframe ->
-            val prevMatchedWireframe = prevSnapshotAsMap.remove(currentWireframe.id())
-            if (prevMatchedWireframe != null) {
-                resolveUpdateMutation(prevMatchedWireframe, currentWireframe)?.let {
-                    updates.add(it)
+        val removes: MutableList<MobileSegment.Remove> = LinkedList()
+        var removalOffsets = Array(oldSnapshot.size) { 0 }
+        var runningOffset = 0
+
+        oa.forEachIndexed { index, entry ->
+            removalOffsets[index] = runningOffset
+            if (entry is Entry.Reference) {
+                // Old element was removed
+                removes.add(MobileSegment.Remove(oldSnapshot[index].id()))
+                runningOffset++
+            }
+        }
+
+        runningOffset = 0
+
+        na.forEachIndexed { index, entry ->
+            when (entry) {
+                is Entry.Index -> {
+                    val indexInOld = entry.index
+                    val removalOffset = removalOffsets[indexInOld]
+                    val newElement = newSnapshot[index]
+                    val oldElement = oldSnapshot[indexInOld]
+                    if ((indexInOld - removalOffset + runningOffset) != index) {
+                        // Old element was moved to another position:
+                        val previousID = if (index > 0) newSnapshot[index - 1].id() else null
```

Comment:
why do we compare index with 0 and take `index-1` later?

Reply:
Hmmm...because if you do not do that check you will have an `IOB` exception.

Reply:
Sorry, I meant what happens with `index=0` (we shouldn't use it?) and why we substract 1 to get an element from `newSnapshot`.

Reply:
we use it, basically we try to get the `previousId` here which in case `index=0` it will be null (see the else branch)